File: src/modules/organization/services/organization-update.service.ts
import {
  ConflictException,
  Injectable,
  NotFoundException,
} from '@nestjs/common';
import { PrismaService } from 'src/infra/database/prisma.service';
import { UserRequest } from 'src/shared/types/user-request';
import { userOrganizationValidator } from 'src/shared/validators/user-organization.validator';
import { OrganizationUpdateDto } from '../domain/dto/organization-update.dto';
import { OrganizationEntity } from '../domain/entities/organization.entity';

@Injectable()
export class OrganizationUpdateService {
  constructor(private prisma: PrismaService) {}
  async execute(
    id: number,
    dto: OrganizationUpdateDto,
    userRequest: UserRequest,
  ): Promise<OrganizationEntity> {
    const organization = await this.prisma.organization.findFirst({
      where: {
        id: id,
        deletedAt: null,
      },
    });

    if (!organization)
      throw new NotFoundException('Organização não encontrada!');

    const existingOrganization = await this.prisma.organization.findFirst({
      where: {
        NOT: {
          id: id,
        },
        OR: [
          {
            name: dto.name,
          },
          {
            email: dto.email,
          },
        ],
        deletedAt: null,
      },
    });

    if (existingOrganization)
      throw new ConflictException('Organização já cadastrada!');

    await userOrganizationValidator(id, userRequest);

    const updatedOrganization = await this.prisma.organization.update({
      data: {
        name: dto.name,
        tradeName: dto.tradeName,
        email: dto.email,
        status: dto.status,
        address: {
          update: dto.address,
        },
      },
      where: {
        id: id,
      },
    });

    return new OrganizationEntity(updatedOrganization);
  }
}

File: src/modules/organization/services/organization_update_service.py
from django.db.models import Q
from src.shared.types.user_request import UserRequest
from src.modules.organization.models import Organization
from src.modules.organization.domain.dto.organization_update_dto import (
    OrganizationUpdateDTO,
)
from src.shared.exceptions.conflict_exception import ConflictException
from src.shared.exceptions.not_found_exception import NotFoundException
from src.shared.validators.user_organization_validator import (
    user_organization_validator,
)


class OrganizationUpdateService:
    @classmethod
    def execute(self, id: int, dto: OrganizationUpdateDTO, user_request: UserRequest):

        organization = Organization.objects.filter(id=id, deleted_at=None).first()

        if organization is None:
            raise NotFoundException("Organização não encontrada!")

        existing_organization = Organization.objects.filter(
            ~Q(id=id) & (Q(name=dto.name) | Q(email=dto.email)) & Q(deleted_at=None)
        ).first()

        if existing_organization is not None:
            raise ConflictException("Organização já cadastrada!")

        user_organization_validator(id, user_request)

        if dto.name is not None:
            organization.name = dto.name
        if dto.trade_name is not None:
            organization.trade_name = dto.trade_name
        if dto.email is not None:
            organization.email = dto.email
        if dto.status is not None:
            organization.status = dto.status

        if dto.address is not None:
            address = organization.address
            if dto.address.state is not None:
                address.state = dto.address.state
            if dto.address.city is not None:
                address.city = dto.address.city
            if dto.address.street is not None:
                address.street = dto.address.street
            if dto.address.neighborhood is not None:
                address.neighborhood = dto.address.neighborhood
            if dto.address.complement is not None:
                address.complement = dto.address.complement

            address.save()

        organization.save()

        return organization
